Rotation manager general docs (#2170)
hashicorp/web-unified-docs: content/vault/v2.x (rc)/content/docs/enterprise/rotation-manager/overview.mdx

---
layout: docs

> page_title: Rotation manager

Reviewer: ```suggestion
page_title: Automated Rotation
```

description: >-
  Vault Enterprise uses a centralized rotation manager to automate credential
  rotation on a scheduled basis with configurable retry behavior

>   and fault-tolerant queue management.

Reviewer: ```suggestion
  Vault Enterprise features a centralized centralized system to automate credential
  rotation on a scheduled basis with configurable retry behavior.
```

---

# Rotation manager

@include 'alerts/enterprise-and-hcp.mdx'

Vault Enterprise includes rotation manager, a centralized system that runs in the background
to automate schedule-based or period-based credential rotation. The rotation manager maintains
a priority queue of credentials, checks for pending rotations every 5 seconds,

> and dispatches rotation requests to the appropriate plugin.

Reviewer: ```suggestion
Vault Enterprise features a centralized system to automate schedule-based 
or period-based credential rotation. Vault periodically evaluates pending rotations,
and dispatches requests to the appropriate plugins.
```

The rotation manager marks a rotation entry as orphaned when it
exceeds the maximum retry limits defined in a rotation policy. Orphaning halts
the rotation attempts and requires manual intervention to resolve the
error. Refer to the [rotation orphaning documentation](/vault/api-docs/system/rotation-orphans)

> for more information.

Reviewer: ```suggestion
## Handling Rotation Failures

If a rotation attempt exceeds the maximum retry limits defined within a rotation policy,
Vault marks the entry as orphaned. This state serves as a safety mechanism:

- Halts Automation: Further automated rotation attempts are paused to prevent loops or service degradation.
- Requires Intervention: Manual troubleshooting is necessary to resolve the root error and restore the credential to a healthy state.

For detailed recovery steps, please refer to the [rotation orphaning documentation](/vault/api-docs/system/rotation-orphans).
```

## High availability

The rotation manager runs only on the active node of a cluster. When leadership
changes, the active node restores rotation state from storage and resumes

> processing the queue.

Reviewer: ```suggestion
Automated rotations execute exclusively on the active node of a cluster. In the event of a leadership change, the new active node automatically restores the rotation state from storage and resumes processing rotations where the previous node left off.
```

For replicated clusters:

- **Shared mounts**: Rotations occur on the active node of the primary cluster.
- **Local mounts**: Rotations occur independently on the active node of each cluster.

## Exponential backoff

> When a rotation fails, the rotation manager requeues the rotation entry using an

Reviewer: ```suggestion
When a rotation fails, the Vault requeues the rotation entry using an
```

exponential backoff schedule to prevent continuous high-frequency retries from
stressing the system. The backoff schedule also includes a randomized factor
(jitter) to help ensure a more even load distribution. The backoff schedule
has 3 elements:

1. **Initial delay** - the first retry attempt uses a minimum delay of 10 seconds.
1. **Doubling interval** - Subsequent delays double the previous backoff interval
(e.g., 10s, 20s, 40s, 80s) and apply jitter to further stagger the retry attempt.
1. **Maximum cap** - Vault caps the delay for any single retry at 5 minutes (300

> seconds).

Reviewer: These details will be helpful for operators that want to understand exactly how retries will work. However, I am wondering if we should also forgo providing exact numbers in the even that we change the implementation?

## Policy-driven

The rotation manager allows for policies to be set that define the retry behavior

> for a given scheduled rotation. Read the [rotation policy documentation](enterprise/rotation-manager/rotation-policies) for more information on how to configure rotation policies and their effect on the rotation manager's retry behavior.

Reviewer: ```suggestion
Vault supports rotation policies to be set that define the retry behavior
for a given scheduled rotation. Read the [rotation policy documentation](enterprise/rotation-manager/rotation-policies) for more information on how to configure rotation policies and their effect on Vault's retry behavior.
```

## Checking a rotation status

Read the role or credential in the plugin to check the rotation status. The response includes

> the fields `last_vault_rotation` which states the timestamp of the last successful rotation

Reviewer: ```suggestion
the fields `last_vault_rotation` which is the timestamp of the last successful rotation
```

performed by Vault. Vault updates the field when an automated rotation succeeds
and whenever a manual rotation occurs between scheduled rotations. The response
also includes `next_vault_rotation`

> which states the timestamp of the next scheduled automatic rotation. The `ttl` field

Reviewer: ```suggestion
which is the timestamp of the next scheduled automatic rotation. The `ttl` field
```

indicates the number of seconds until the next scheduled rotation.

For example, reading an LDAP static role credential includes the following information:

```shell-session
$ vault read ldap/static-cred/my-role
```

```
Key					           Value
---					           -----
last_vault_rotation    2026-04-02T23:30:00Z
next_vault_rotation    2026-04-03T11:30:00Z
ttl                    84600
...
```

## Troubleshooting

> Rotation manager logs rotation activity to the Vault server logs such as successful rotations,

Reviewer: ```suggestion
Vault logs rotation activity to the Vault server logs such as successful rotations,
```

rotation failures, retry attempts, and orphaning. Vault includes metadata in the log statements
to identify the credential being rotated and error information in the case of a failure. For example:

```
[INFO]  rotation-job-manager: successfully rotated job: rotationID=database/config/my-database
[ERROR] rotation-job-manager: rotation failed, attempting to re-queue: rotationID=database/config/my-database error="connection refused"
[ERROR] rotation-job-manager: orphaning item, please re-register after resolving the error: rotationID=database/config/my-database
```

## Migration from legacy rotation system in plugins

<Warning>

When you unseal Vault the first time after making an upgrade that includes a plugin onboarding
to the rotation manager system, Vault must migrate preexisting rotation entries into the new queueing

> system. 

Reviewer: ```suggestion
While some plugins were originally designed with independent rotation systems, Vault updates may transition them to the centralized framework. When you unseal Vault for the first time following an upgrade that enables this transition for a plugin, Vault automatically migrates all preexisting rotation entries into the new system.
```
